### PR TITLE
Throw Error dialog on missing gstreamer plugin

### PR DESCRIPTION
**Pull request by arshubham**
### **WIP**

Fixes https://github.com/elementary/videos/issues/117 , Fixes https://github.com/elementary/videos/issues/102 , Fixes https://github.com/elementary/videos/issues/97

To test try playing a flv file in Videos. Make sure `gstreamer1.0-plugins-bad` is uninstalled.
image

Test file: https://send.firefox.com/download/f1130affe90869e5/#L1v8Xtn2BxBBCiqSBaunug

---
This should ideally work however I am getting following error from packagekit when `install_plugins_sync` is called with the parameter `gstreamer|1.0|io.elementary.videos|Flash demuxer|decoder-video/x-flv` The code is based on music's implementation https://github.com/elementary/music/blob/master/src/Dialogs/InstallGstreamerPluginsDialog.vala

```
** Message: 21:31:13.259: PackageKit: xid = 0
** Message: 21:31:13.259: PackageKit: desktop_id = (null)
** Message: 21:31:13.260: PackageKit: Codec nice name: Flash demuxer
** Message: 21:31:13.260: PackageKit: structure: gstreamer1(decoder-video/x-flv)()(64bit)
** Message: 21:31:13.262: PackageKit: Did not install codec: GDBus.Error:org.freedesktop.DBus.Error.ServiceUnknown: The name org.freedesktop.PackageKit was not provided by any .service files
```

### CODE REVIEW

**Comment by arshubham** [comment]
Experiencing a cluttergst crash. Insvestigating

**Comment by 4jNsY6fCVqZv** [comment]
@arshubham Thanks for working so hard on Videos lately!

When I look at your screenshot in this issue, I ask myself: If I can't play .flv files because I haven't installed gstreamer1.0-plugins-bad anymore, what's the point of the "Play Anyway" button? I can press it, but Videos already know that I can't play the file.
Wouldn't it be clever, if Videos check if the file can be played and if no, that I'm offered to install the missing plugin? Or otherwise abort the playback.

Or did I just miss something?

**Comment by arshubham** [comment]
@4jNsY6fCVqZv Yes that is the plan. Still looking for a good way for installation of plugins. This Pr is just the part 1 for the enhancement. For now play anyway just preserves the current functionality. Opened this for initial feedback. 

> what's the point of the "Play Anyway" button

In current behavior, if the plugin is missing. Sometimes the Videos will  just seek without any audio or video.

**Comment by 4jNsY6fCVqZv** [comment]
> Yes that is the plan.

Sounds wonderful, thank you!

> sometimes the Videos will just seek without any audio or video.

And that makes sense that Videos try that? If a video has no video, it could be a radio play/audio book. So it's just audio and more of a case for playback with Music? If a video doesn't have audio, it's maybe a silent movie. If neither is the case, in both cases additional plugins are needed to play the video. Wouldn't it be clever, if Videos check, for which functionality the missing plugin would be responsible and then gives appropriate tips? 
For example the plugin for sound playback is missing, Videos says: "Hej, you are missing the plugin for sound playback, if it is a movie without sound just press 'play anyway', otherwise you can install the missing plugin here. The same applies to the example if video support is missing.
What do you think?

**Comment by 4jNsY6fCVqZv** [comment]
> Yes that is the plan.

Sounds wonderful, thank you!

> sometimes the Videos will just seek without any audio or video.

And that makes sense that Videos try that? If a video has no video, it could be a radio play/audio book. So it's just audio and more of a case for playback with Music? If a video doesn't have audio, it's maybe a silent movie. If neither is the case, in both cases additional plugins are needed to play the video. Wouldn't it be clever, if Videos check, for which functionality the missing plugin would be responsible and then gives appropriate tips? 
For example the plugin for sound playback is missing, Videos says: "Hey, you are missing the plugin for sound playback, if it is a movie without sound just press 'play anyway', otherwise you can install the missing plugin here. The same applies to the example if video support is missing.
What do you think?

File: src/Widgets/Player/PlayerPage.vala [hunk 1]
```diff
@@ -269,8 +271,9 @@ namespace Audience {
             debug ("Opening %s", uri);
 
             var file = File.new_for_uri (uri);
+            FileInfo info = new FileInfo ();
```
Comment (tintou) [inline src/Widgets/Player/PlayerPage.vala]:
```suggestion
            FileInfo? info = null;
```

File: src/Widgets/Player/PlayerPage.vala [hunk 2]
```diff
@@ -293,12 +296,41 @@ namespace Audience {
...
+                var message = Gst.PbUtils.missing_decoder_message_new (pipeline, caps);
+                var installer = Gst.PbUtils.missing_plugin_message_get_installer_detail (message);
+                var plugin_name = Gst.PbUtils.missing_plugin_message_get_description (message);
+                var context = new Gst.PbUtils.InstallPluginsContext ();
+
+                var missing_plugin_dialog = new MissingPluginDialog (uri, info.get_name (), plugin_name);
+                missing_plugin_dialog.present ();
+                missing_plugin_dialog.response.connect (type => {
+                    if (type == Gtk.ResponseType.ACCEPT) {
+                        Gst.PbUtils.InstallPluginsReturn ret = Gst.PbUtils.install_plugins_sync ({ installer }, context);
```
Comment (tintou) [inline src/Widgets/Player/PlayerPage.vala]: Use `install_plugins_async` to not freeze the UI during lookup

**Comment by 4jNsY6fCVqZv** [comment]
@arshubham Thank you, that's a good improvement for me!
Is "Details" expanded by default? Otherwise users won't see which plugin is missing or should be installed.
Btw. is there also a use case where several plugins have to be installed at once to be able to play a file?

**Comment by 4jNsY6fCVqZv** [comment]
@arshubham Thanks, looks like a good improvement to me already! Is "Details" expanded by default? Otherwise users won't see which plugin is missing or should be installed.
Is there also a use case where several plugins have to be installed at once to be able to play a file?

**Comment by arshubham** [comment]
@4jNsY6fCVqZv Details isn't expanded by default right now. However plan here is to more this dialog into appcenter itself, more improvements can be done there. 
Currently this check happens on playing a file, so I don't think there will be a case where multiple plugins are required to be installed. Gstreamer should handle that anyway.

**Comment by 4jNsY6fCVqZv** [comment]
> However plan here is to move this dialog into appcenter itself, more improvements can be done there.

How could this integration into AppCenter look like?

**Comment by danrabbit** [comment]
@arshubham Not sure if you want to continue working on this branch? Looks like it has conflicts with master and you have some review feedback from tintou

**Comment by jeremypw** [comment]
@danrabbit Looks like the author has been inactive on GitHub since the beginning of March and not on this PR last since October - could well have more important things to deal with at the moment.  If you like I can try and get it in shape.

**Comment by davidmhewitt** [comment]
I've fixed this up now, requires some work in AppCenter to provide the necessary D-Bus interfaces for installing the required plugins, as per: https://github.com/elementary/appcenter/issues/623

I've assigned that to myself and will work on that next.

**Pull request closed without merging**